disco_aws_automation/disco_deploy.py

```diff
@@ -390,6 +391,105 @@ def handle_tested_ami(self, old_hostclass_dict, ami, desired_size,
         self._disco_aws.terminate(self._get_new_instances(ami.id), use_autoscaling=True)
         self._disco_aws.spinup([post_hostclass_dict])
 
+    # Disable too many local variables, branches, and statements because this method handles blue/green from
+    # end to end.
+    # pylint: disable=too-many-locals,too-many-branches,too-many-statements
+    def handle_blue_green_ami(self, pipeline_dict, ami, old_group,
+                              deployable=False, run_tests=False, dry_run=False):
+        '''
+        Tests hostclasses which we can deploy normally
+
+        Deploys AMIs into a new autoscaling group. If the new AMI passes tests, the old ASG is destroyed,
+        and the new ASG is preserved. Otherwise, the original ASG is preserved.
+
+        Also creates a separate testing ELB that is used for the purposes of integration tests.
+        '''
+        logging.info("testing %s hostclass %s AMI %s with %s deployment strategy",
+                     "deployable" if deployable else "non-deployable", pipeline_dict["hostclass"], ami.id,
+                     DEPLOYMENT_STRATEGY_BLUE_GREEN)
+
+        if dry_run:
+            return
+
+        hostclass = pipeline_dict["hostclass"]
+        uses_elb = is_truthy(self.hostclass_option_default(hostclass, "elb", "no"))
+
+        new_config = copy.deepcopy(pipeline_dict)
+        new_config["sequence"] = 1
+        new_config["smoke_test"] = "no"
+        new_config["ami"] = ami.id
+        new_config["max_size"] = pipeline_dict["max_size"] if "max_size" in pipeline_dict else 1
+        new_config["min_size"] = pipeline_dict["min_size"] if "min_size" in pipeline_dict else 1
+        new_config["desired_size"] = pipeline_dict["desired_size"] if "desired_size" in pipeline_dict else 1
+
+        # Spinup our new autoscaling group in testing mode, making one even if one already exists.
+        self._disco_aws.spinup([new_config], create_if_exists=True, testing=True)
```

Comment:
Before we spinup a new AG, could we make sure there currently exists no more than one already? I know it won't solve all the race situations, we'll do our best I guess.